### PR TITLE
Replace deprecated tbb::task_cheduler_init for tbb::task_arena

### PR DESCRIPTION
**Pull request by xvallspl**
Replacing tbb::task_scheduler_init for tbb::task_arena.  It does not change current ROOT MT behaviour.

Description in the commit message:

>     replace TPoolManager by RTaskArenaWrapper
> 
>     Replace all uses of TPoolManager, based on the deprecated
>     tbb_task_scheduler_init and its implicit task_arena, by explicit
>     manipulation of our own central instance of task_arena, as
>     suggested by intel:
>     https://software.intel.com/sites/default/files/managed/b2/d2/TBBRevamp.pdf
> 
>     Work out some gymnastics to keep tbb out of the headers,
>     encapsulating it within the new RTaskArenaWrapper class. Couldn't
>     find a solution for the forward-declaration tbb::task_arena, which
>     is an alias of a versioned namespace dependent class that we are
>     forced to forward declare instead (tbb::interface7::tbb_task_arena).

### CODE REVIEW

File: core/imt/src/RTaskArena.hxx [hunk 1]
```diff
@@ -0,0 +1,143 @@
...
+//references to the previous one are gone and the object destroyed.
+////////////////////////////////////////////////////////////////////////////////
+static std::shared_ptr<ROOT::Internal::RTaskArenaWrapper> GetGlobalTaskArena()
+{
+   static std::weak_ptr<ROOT::Internal::RTaskArenaWrapper> weak_GTAWrapper;
+
+   if (weak_GTAWrapper.expired()) {
+      std::shared_ptr<ROOT::Internal::RTaskArenaWrapper> shared_GTAWrapper(new ROOT::Internal::RTaskArenaWrapper());
+      weak_GTAWrapper = shared_GTAWrapper;
+      return weak_GTAWrapper.lock();
```
Comment (hageboeck) [inline core/imt/src/RTaskArena.hxx:103]:
```suggestion
      return shared_GTAWrapper;
```

Comment (xvallspl) [reply core/imt/src/RTaskArena.hxx:103]: Why? It's the same and it's more coherent with what I return on the other case. Makes it easier to understand.

File: core/imt/src/RTaskArena.hxx [hunk 2]
```diff
@@ -0,0 +1,143 @@
...
+         std::string periodFile("/sys/fs/cgroup/cpuacct/cpu.cfs_period_us");
+         f.open(periodFile);
+         float cfs_period;
+         f>>cfs_period;
+         f.close();
+         return static_cast<int>(std::ceil(cfs_quota/cfs_period));
+      }
+   }
+#endif
+   return std::thread::hardware_concurrency();
```
[the remark this answers is not in the document]
Comment (xvallspl) [reply core/imt/src/RTaskArena.hxx:86]: but you need to create a task_arena. What i do is the same, but in in InitializeArena. There I ask to my already instantiated but not initialized arena

File: core/imt/src/RTaskArena.hxx [hunk 3]
```diff
@@ -0,0 +1,143 @@
...
+}
+
+////////////////////////////////////////////////////////////////////////////////
+// Factory function returning a shared pointer to the instance of the global
+//RTaskArenaWrapper.
+//
+// Allows for reinstantiation of the global RTaskArenaWrapper once all the
+//references to the previous one are gone and the object destroyed.
+////////////////////////////////////////////////////////////////////////////////
+static std::shared_ptr<ROOT::Internal::RTaskArenaWrapper> GetGlobalTaskArena()
```
[the remark this answers is not in the document]
Comment (xvallspl) [reply core/imt/src/RTaskArena.hxx:96]: Way of replicating previous behaviour, as ROOT::GetPoolSize() might be asked when the task arena hasn't been initialized.

In any case, this header is in core/imt/src/ and not installed

[the remark this answers is not in the document]
Comment (eguiraud) [reply core/imt/src/RTaskArena.hxx:96]: So I guess it comes down to: is there any way we can implement `GetThreadPoolSize` without a method that retrieves an uninitialized task arena?

File: core/imt/src/TThreadExecutor.cxx [hunk 4]
```diff
@@ -139,37 +136,33 @@ namespace ROOT {
    ///
    /// At construction time, TThreadExecutor automatically enables ROOT's thread-safety locks as per calling
    /// ROOT::EnableThreadSafety().
-   TThreadExecutor::TThreadExecutor(UInt_t nThreads)
-   {
-      ROOT::EnableThreadSafety();
```
Comment (eguiraud) [inline core/imt/src/TThreadExecutor.cxx:144]: why did you remove the call to `EnableThreadSafety()`?

Comment (eguiraud) [reply core/imt/src/TThreadExecutor.cxx:144]: nevermind, `InitGlobalTaskArena` calls it now

File: core/imt/src/TThreadExecutor.cxx [hunk 5]
```diff
@@ -139,37 +136,33 @@ namespace ROOT {
...
    {
-      return ROOT::Internal::ParallelReduceHelper<float>(objs, redfunc);
+      return fTaskArenaW->Execute([&]{
+               return ROOT::Internal::ParallelReduceHelper<float>(objs, redfunc);
+      });
    }
 
    unsigned TThreadExecutor::GetPoolSize(){
-      return ROOT::Internal::TPoolManager::GetPoolSize();
+      return  static_cast<UInt_t>(fTaskArenaW->TaskArenaSize());
```
Comment (eguiraud) [inline core/imt/src/TThreadExecutor.cxx:165]: nitpick: should be `static_const<unsigned>` just for consistency

File: core/imt/src/TThreadExecutor.cxx (hunk 4 above)
Comment (xvallspl) [reply core/imt/src/TThreadExecutor.cxx:144]: I moved it to the task_arena initialization

**Comment by xvallspl** [comment]
Still debugging this

**Comment by xvallspl** [comment]
I give up, I'll forward declare tbb::interface7::task_arena and ask the TBB developers about the chances for the interface-dependent name to change in the near future

**Comment by xvallspl** [comment]
Will probably move the initialization of the task arena into RTaskArenaWrapper next

**Comment by xvallspl** [comment]
Also, enabling thread safety on initialization of the task arena should allow us to get rid of all ROOT::EnableThreadSafety() calls throughout ROOT

File: core/imt/src/RTaskArena.cxx [hunk 6]
```diff
@@ -0,0 +1,146 @@
...
+std::shared_ptr<ROOT::Internal::RTaskArenaWrapper> GetGlobalTaskArena(unsigned maxConcurrency)
+{
+   static std::weak_ptr<ROOT::Internal::RTaskArenaWrapper> weak_GTAWrapper;
+   if (weak_GTAWrapper.expired()) {
+      std::shared_ptr<ROOT::Internal::RTaskArenaWrapper> shared_GTAWrapper(new ROOT::Internal::RTaskArenaWrapper(maxConcurrency));
+      weak_GTAWrapper = shared_GTAWrapper;
+      return weak_GTAWrapper.lock();
+   }
+
+   return weak_GTAWrapper.lock();
```
Comment (bernhardmgruber) [inline core/imt/src/RTaskArena.cxx:142]: Potential race condition. The the lifetime of the shared pointer pointed to by `weak_GTAWrapper` could expire between the call to `expired()` and the last `lock()`. Try to lock directly and inspect the result:

```
std::shared_ptr<ROOT::Internal::RTaskArenaWrapper> GetGlobalTaskArena(unsigned maxConcurrency)
{
  static std::weak_ptr<ROOT::Internal::RTaskArenaWrapper> weak_GTAWrapper;
  if (auto sp = weak_GTAWrapper.lock())
    return sp;
  auto sp = std::make_shared<ROOT::Internal::RTaskArenaWrapper>(maxConcurrency);
  weak_GTAWrapper = sp;
  return sp;
}
```
And prefer `std::make_shared` instead of repeating yourself.

File: core/imt/inc/ROOT/RTaskArena.hxx [hunk 7]
```diff
@@ -0,0 +1,79 @@
...
+///
+/// Necessary in order to keep tbb away from ROOT headers.
+/// This class is thought out to be used as a singleton.
+////////////////////////////////////////////////////////////////////////////////
+class RTaskArenaWrapper {
+public:
+   RTaskArenaWrapper(unsigned maxConcurrency = 0);
+   ~RTaskArenaWrapper(); // necessary to set size back to zero
+   static unsigned TaskArenaSize();
+   std::unique_ptr<tbb::task_arena> &Access();
```
[the remark this answers is not in the document]
Comment (xvallspl) [reply core/imt/inc/ROOT/RTaskArena.hxx:58]: Because tbb::task_arena is forward declared, and I would need to know it's size to allocate the data member.

File: core/imt/src/RTaskArena.cxx (hunk 6 above)
Comment (xvallspl) [reply core/imt/src/RTaskArena.cxx:142]: Yeah, this one we knew and decided to leave as is for the minimal chance of that happening You will have to call it concurrently from at least two different threads explicitly managed by the user, assuming you didn't initialize any ROOT MT class before.

File: core/imt/inc/ROOT/RTaskArena.hxx [hunk 8]
```diff
@@ -0,0 +1,79 @@
...
+#  error "Cannot use ROOT::Internal::RTaskArenaWrapper without defining R__USE_IMT."
+# endif
+#else
+
+/// tbb::task_arena is an alias of tbb::interface7::task_arena, which doesn't allow
+/// to forward declare tbb::task_arena without forward declaring tbb::interface7
+namespace tbb{
+namespace interface7{class task_arena;}
+using task_arena = interface7::task_arena;
+}
```
[the remark this answers is not in the document]
Comment (xvallspl) [reply core/imt/inc/ROOT/RTaskArena.hxx:42]: Yes, that was discussed in the previous PPP, it's in the commit message, and I tried to solve it in two different PRs unsuccessfully. Didn't find a way to keep at least a forward declaration of tbb::task_arena out of the headers because we need tbb::task_arena::execute() in TThreadExecutor.cxx, which implies the definition of the method.

On the other hand, tbb::task_arena is an alias to tbb::interface7::task_arena and  you cannot forward declare aliases https://stackoverflow.com/questions/18834645/c-forward-declaration-and-alias-with-using-or-typedef

File: core/imt/inc/ROOT/RTaskArena.hxx (hunk 7 above)
[the remark this answers is not in the document]
Comment (bernhardmgruber) [reply core/imt/inc/ROOT/RTaskArena.hxx:58]: `tbb::task_arena& Access();` should compile perfectly fine with `tbb::task_arena` just being forward declared. Even `tbb::task_arena Access();` should compile.

File: core/imt/src/RTaskArena.cxx (hunk 6 above)
Comment (bernhardmgruber) [reply core/imt/src/RTaskArena.cxx:142]: Are you serious? This is a potential bug and the fix is easy. There is really no reason to not apply it. It takes you 5min to do so and prevent a potential 3h or more debugging session of someone in the future.

File: core/imt/inc/ROOT/RTaskArena.hxx (hunk 8 above)
[the remark this answers is not in the document]
Comment (bernhardmgruber) [reply core/imt/inc/ROOT/RTaskArena.hxx:42]: I would just #include the corresponding TBB header and have a smoother TBB upgrade next time or allowing people to use different versions of TBB with ROOT. But as you said, it was discussed in the PPP meeting.

[the remark this answers is not in the document]
Comment (xvallspl) [reply core/imt/inc/ROOT/RTaskArena.hxx:42]: I understand, but  the requirement here was keeping TBB out of ROOT headers

File: core/imt/src/RTaskArena.cxx [hunk 9]
```diff
@@ -0,0 +1,146 @@
...
+////////////////////////////////////////////////////////////////////////////////
+static Int_t LogicalCPUBandwithControl()
+{
+#ifdef R__LINUX
+   // Check for CFS bandwith control
+   std::ifstream f;
+   std::string quotaFile("/sys/fs/cgroup/cpuacct/cpu.cfs_quota_us");
+   struct stat buffer;
+   // Does the file exist?
+   if(stat(quotaFile.c_str(), &buffer) == 0) {
```
[the remark this answers is not in the document]
Comment (hageboeck) [reply core/imt/src/RTaskArena.cxx:60]: Neither boost nor c++17 can be used here, IIUC. But creating the stream and checking it should work.

File: core/imt/inc/ROOT/RTaskArena.hxx (hunk 7 above)
[the remark this answers is not in the document]
Comment (xvallspl) [reply core/imt/inc/ROOT/RTaskArena.hxx:58]: Oh, of course, sorry. You have a point. Thought you were commenting on the data member :)

File: core/imt/src/RTaskArena.cxx (hunk 6 above)
Comment (xvallspl) [reply core/imt/src/RTaskArena.cxx:136]: Sorry, I didn't look properly at your code. I thought we were talking about adding std::locks and making it a critical section. Nice solution, didn't think of that.

File: core/imt/inc/ROOT/RTaskArena.hxx [hunk 10]
```diff
@@ -0,0 +1,79 @@
...
+////////////////////////////////////////////////////////////////////////////////
+class RTaskArenaWrapper {
+public:
+   RTaskArenaWrapper(unsigned maxConcurrency = 0);
+   ~RTaskArenaWrapper(); // necessary to set size back to zero
+   static unsigned TaskArenaSize();
+   std::unique_ptr<tbb::task_arena> &Access();
+private:
+   std::unique_ptr<tbb::task_arena> fTBBArena;
+   static unsigned fNWorkers;
```
[the remark this answers is not in the document]
Comment (xvallspl) [reply core/imt/inc/ROOT/RTaskArena.hxx:61]: I see it conceptually as a member of the class.

File: core/imt/src/RTaskArena.cxx [hunk 11]
```diff
@@ -123,17 +123,14 @@ tbb::task_arena &RTaskArenaWrapper::Access()
...
-      weak_GTAWrapper = shared_GTAWrapper;
-      return weak_GTAWrapper.lock();
-   }
-
-   return weak_GTAWrapper.lock();
+   if (auto sp = weak_GTAWrapper.lock())
+      return sp;
+   auto sp = std::make_shared<ROOT::Internal::RTaskArenaWrapper>(maxConcurrency);
+   weak_GTAWrapper = sp;
+   return sp;
```
Comment (hageboeck) [inline core/imt/src/RTaskArena.cxx:133]: I assume you can still have a race condition. If two threads try to get the task arena, and it doesn't exist, both will get a `nullptr`. Now, they both make a shared pointer to arenas. One arena will be discarded once the returned pointer goes out of scope.

Comment (bernhardmgruber) [reply core/imt/src/RTaskArena.cxx:133]: You are right! And even further, two threads could concurrently assign to `weak_GTAWrapper` which is also a race condition.

Now I ask myself what the semantic of this function should be? It currently seems like you want to destroy the task arena, once the last user is done with it. Why is this even needed? Why can be not just keep it around?

Like:
```
ROOT::Internal::RTaskArenaWrapper& GetGlobalTaskArena(unsigned maxConcurrency) {
  static ROOT::Internal::RTaskArenaWrapper taw{maxConcurrency};
  return taw;
}
```

I furthermore ask myself what the meaning of `maxConcurrency` is for the caller. Because depending on whether there is any other thread still keeping the task arena alive with a different value of `maxConcurrency` I ether get a task arena with the `maxConcurrency` I requested or a different one.

Comment (xvallspl) [reply core/imt/src/RTaskArena.cxx:133]: True. So...you guys want me to put locks in there?

Comment (bernhardmgruber) [reply core/imt/src/RTaskArena.cxx:133]: No. Not now. First please clarify what are the semantics of this function and the lifetime of the task arena.

Comment (eguiraud) [reply core/imt/src/RTaskArena.cxx:133]: About the race condition: iiuc you want to solve [ROOT-10762](https://sft.its.cern.ch/jira/browse/ROOT-10762) here, which I think is more of a nice-to-have than a must

Comment (hageboeck) [reply core/imt/src/RTaskArena.cxx:133]: A function-local static would be thread safe, and live till the end of ROOT. Sounds ok to me.

Comment (bernhardmgruber) [reply core/imt/src/RTaskArena.cxx:133]: @eguiraud I do not want to solve any already existing issues. Just not create new ones ;)
@hageboeck If having a single task arena for the entire lifetime of ROOT is not wrong, than I guess the function local static is a safe bet.

Comment (eguiraud) [reply core/imt/src/RTaskArena.cxx:133]: if i'm not mistaken, originally, the TPoolManager needed to be destructed and reconstructed to change the number of threads in the pool, so it could not have static lifetime. Xavi, am I right? If yes, could the new logic accommodate of working threads in a different way?

Comment (hageboeck) [reply core/imt/src/RTaskArena.cxx:133]: Regarding the concurrency, I think we agreed that you can choose it once, and then it's fixed. We would need a mechanism to tell users, though, that if they input another value, it's getting ignored. E.g.
```
EnableIMT(2);
// do stuff
DisableIMT();
EnableIMT(4);
```

Comment (eguiraud) [reply core/imt/src/RTaskArena.cxx:133]: this is not the current behavior and i don't remember agreeing it should be..?

Comment (xvallspl) [reply core/imt/src/RTaskArena.cxx:133]: @bernhardmgruber Oh, but this issue is there already, currently in TPoolManager.

A single task arena for the entire lifetime of ROOT is **not** what we want. And the only way I see to solve the issue right now is to have a critical section.

The idea of this function is to have either none or a single global task_arena alive at any point in time within ROOT.

Right now, this function is called in two cases: When we Initialize a TThreadExecutor and when we call EnableImplicitMT. Both of them hold a shared ptr to the RTaskArenaWrapper during their lifetime, and the RTaskArenaWrapper should be destroyed when every shared owner ceases to exist (TThreadExecutor is destroyed) or releases it (DisableImplicitMT() releases the shared_ptr acquired by EnableImplicitMT). This allows for subsequent reinstantiation of the task_arena, potentially with a different number of worker threads.

Comment (xvallspl) [reply core/imt/src/RTaskArena.cxx:133]: > Regarding the concurrency, I think we agreed that you can choose it once, and then it's fixed. We would need a mechanism to tell users, though, that if they input another value, it's getting ignored. E.g.
> 
> ```
> EnableIMT(2);
> // do stuff
> DisableIMT();
> EnableIMT(4);
> ```

as @eguiraud said, that is not the agreed current behaviour. But the mechanism to warn users when we ignore values is in the PR (and currently in TThreadExecutor but not IMT in master)

Comment (bernhardmgruber) [reply core/imt/src/RTaskArena.cxx:133]: @xvallspl thx for the clarification. Then the task arena has shared ownership/lifetime and a shared_ptr might be a good choice here. And as you pointed out, you might need a mutex here to protect the creation of a new task arena:
```
std::shared_ptr<ROOT::Internal::RTaskArenaWrapper> GetGlobalTaskArena(unsigned maxConcurrency)
{
  static std::weak_ptr<ROOT::Internal::RTaskArenaWrapper> weak_GTAWrapper;
  if (auto sp = weak_GTAWrapper.lock())
    return sp;
  static std::mutex m;
  std::lock_guard _{m};
  if (auto sp = weak_GTAWrapper.lock())
    return sp;
  auto sp = std::make_shared<ROOT::Internal::RTaskArenaWrapper>(maxConcurrency);
  weak_GTAWrapper = sp;
  return sp;
}
```

Comment (xvallspl) [reply core/imt/src/RTaskArena.cxx:133]: @eguiraud ok for this as well? nice-to have is enough. I don't expect anyone to start spamming enough TThreadExecutors concurrently for this to be a performance issue

**Comment by xvallspl** [comment]
@phsft-bot build!

File: core/imt/src/RTaskArena.cxx (hunk 11 above)
Comment (eguiraud) [reply core/imt/src/RTaskArena.cxx:137]: yep, creating TThreadExecutors in a parallel hot loop is not a usecase we care to optimize

File: core/imt/inc/ROOT/RTaskArena.hxx [hunk 12]
```diff
@@ -0,0 +1,79 @@
+// @(#)root/thread:$Id$
```
Comment (hageboeck) [inline core/imt/inc/ROOT/RTaskArena.hxx:1]: Does this line do anything or is it auto-generated by your editor?

File: core/imt/inc/ROOT/RTaskArena.hxx [hunk 13]
```diff
@@ -0,0 +1,79 @@
...
+
+
+////////////////////////////////////////////////////////////////////////////////
+// Factory function returning a shared pointer to the instance of the global
+// RTaskArenaWrapper. The task arena still needs to be initialized with a
+// call to RTaskArena::Initialize(maxConcurrency)
+//
+// Allows for reinstantiation of the global RTaskArenaWrapper once all the
+// references to the previous one are gone and the object destroyed.
+////////////////////////////////////////////////////////////////////////////////
```
Comment (hageboeck) [inline core/imt/inc/ROOT/RTaskArena.hxx:72]: Make it doxygen readable?
```suggestion
/// Factory function returning a shared pointer to the instance of the global
/// RTaskArenaWrapper. The task arena still needs to be initialized with a
/// call to RTaskArena::Initialize(maxConcurrency)
///
/// Allows for reinstantiation of the global RTaskArenaWrapper once all the
/// references to the previous one are gone and the object destroyed.
////////////////////////////////////////////////////////////////////////////////
```

File: core/imt/inc/ROOT/RTaskArena.hxx [hunk 14]
```diff
@@ -0,0 +1,79 @@
...
+}
+
+namespace ROOT {
+namespace Internal {
+
+////////////////////////////////////////////////////////////////////////////////
+/// Wrapper for tbb::task_arena.
+///
+/// Necessary in order to keep tbb away from ROOT headers.
+/// This class is thought out to be used as a singleton.
```
Comment (hageboeck) [inline core/imt/inc/ROOT/RTaskArena.hxx:51]: So should it be a singleton, then? I know that the singleton is out of fashion, so maybe not mention "Singleton" here and say how to instantiate one?
I guess it shouldn't be a singleton. Maybe people want to have their second task arena for their framework, and probably that would work.

File: core/imt/inc/ROOT/RTaskArena.hxx (hunk 12 above)
Comment (xvallspl) [reply core/imt/inc/ROOT/RTaskArena.hxx:1]: This is in all the other imt headers, I just copied the copyright statement adapting the author and the date. Not really sure of what it does or if It's necessary 🤷‍♂️

**Comment by xvallspl** [comment]
> [3fa8acb](https://github.com/root-project/root/commit/3fa8acb3c445f2949ea73ef62c89d761f150ebb9) has a funny character int he commit message.

Will correct as soon as we are done building

File: core/imt/inc/ROOT/RTaskArena.hxx (hunk 12 above)
Comment (eguiraud) [reply core/imt/inc/ROOT/RTaskArena.hxx:1]: you can ask Axel, I think it was used by the old homegrown reference guide generator. I don't think it's a doxygen thing, and it could be removed

File: core/imt/src/RTaskArena.cxx [hunk 15]
```diff
@@ -0,0 +1,141 @@
...
+#endif
+   return std::thread::hardware_concurrency();
+}
+
+namespace ROOT{
+namespace Internal {
+
+
+////////////////////////////////////////////////////////////////////////////////
+/// Initializes the tbb::task_arena within RTaskArenaWrapper
```
Comment (hageboeck) [inline core/imt/src/RTaskArena.cxx:78]:
```suggestion
/// Initializes the tbb::task_arena within RTaskArenaWrapper.
```
Always include the dot, so doxygen generates the correct "brief"

File: core/imt/src/RTaskArena.cxx (hunk 11 above)
Comment (hageboeck) [reply core/imt/src/RTaskArena.cxx:137]: Looks to me like this will work.

**Comment by eguiraud** [comment]
@vgvassilev any idea how to get a travis check on this PR?

**Comment by vgvassilev** [comment]
> @vgvassilev any idea how to get a travis check on this PR?

I have no idea why it is dead for this PR...

File: core/imt/src/RTaskArena.cxx [hunk 16]
```diff
@@ -0,0 +1,141 @@
...
+///
+/// * Can't be reinitialized
+/// * Checks for CPU bandwidth control and avoids oversubscribing
+/// * If no BC in place and maxConcurrency<1, defaults to the default tbb number of threads,
+/// which is CPU affinity aware
+////////////////////////////////////////////////////////////////////////////////
+RTaskArenaWrapper::RTaskArenaWrapper(unsigned maxConcurrency): fTBBArena(new tbb::task_arena{})
+{
+   if (!fTBBArena->is_active()) {
```
[the remark this answers is not in the document]
Comment (xvallspl) [reply core/imt/src/RTaskArena.cxx:87]: The tbb::task_arena is not initialized on construction, you require a call to initialize

[the remark this answers is not in the document]
Comment (eguiraud) [reply core/imt/src/RTaskArena.cxx:87]: What I'm missing is how can this logic ever take a different path then:

* singleton RTaskArenaWrapper is constructed
* also constructs a new fTBBArena
* new fTBBArena is_active returns false
* new fTBBArena is initialized

File: core/imt/src/TImplicitMT.cxx [hunk 17]
```diff
@@ -49,10 +48,7 @@ static std::atomic_int &GetParBranchProcessingCount()
 extern "C" void ROOT_TImplicitMT_EnableImplicitMT(UInt_t numthreads)
```
[the remark this answers is not in the document]
Comment (xvallspl) [reply core/imt/src/TImplicitMT.cxx:48]: I think it had to do with TImplicitMT.cxx being in core/imt and EnableImplicitMT being part of TROOT.h, in core

File: core/imt/test/testRTaskArena.cxx [hunk 18]
```diff
@@ -0,0 +1,168 @@
...
+         f.open("/sys/fs/cgroup/cpuacct/cpu.cfs_period_us"); // period file
+         float cfs_period;
+         f>>cfs_period;
+         f.close();
+         return static_cast<int>(std::ceil(cfs_quota/cfs_period));
+      }
+   }
+#endif
+   return std::thread::hardware_concurrency();
+}
```
[the remark this answers is not in the document]
Comment (xvallspl) [reply core/imt/test/testRTaskArena.cxx:32]: We could, I was not sure if that's what we wanted given that it's a function with a single use in ROOT.

[the remark this answers is not in the document]
Comment (eguiraud) [reply core/imt/test/testRTaskArena.cxx:32]: two uses if you count this one :D I don't think it can be harmful to put it in ROOT::Internal

File: core/imt/src/TImplicitMT.cxx (hunk 17 above)
[the remark this answers is not in the document]
Comment (eguiraud) [reply core/imt/src/TImplicitMT.cxx:48]: we jump through manual symbol loading hoops just because we got the placement of `EnableImplicitMT`? :fearful:

File: core/imt/test/testRTaskArena.cxx [hunk 19]
```diff
@@ -0,0 +1,168 @@
...
+   nCores = plausibleNCores(randGenerator);
+   auto gTAInstance = ROOT::Internal::GetGlobalTaskArena(nCores);
+   ASSERT_EQ(ROOT::Internal::RTaskArenaWrapper::TaskArenaSize(), nCores);
+}
+
+TEST(RTaskArena, SingleInstance)
+{
+   const unsigned nCores  = plausibleNCores(randGenerator);
+   auto gTAInstance1 = ROOT::Internal::GetGlobalTaskArena(nCores);
+   auto gTAInstance2 = ROOT::Internal::GetGlobalTaskArena(plausibleNCores(randGenerator));
```
[the remark this answers is not in the document]
Comment (xvallspl) [reply core/imt/test/testRTaskArena.cxx:78]: Ah, true. Missed that. After moving the RTaskArena initialization function into the constructor now I never trigger the warning in it. I should move the warning to GetGlobalTaskArena

File: core/imt/src/RTaskArena.cxx [hunk 20]
```diff
@@ -0,0 +1,141 @@
...
+///     active in any ROOT Process
+///   - Solves multiple undefined behaviors. Guaranteeing that all classes
+///    use the same task arena avoids interferences and undefined behavior
+///    by providing a single instance of the tbb::task_arena and automated
+///    bookkeeping, instantiation and destruction.
+///
+/// #### Examples:
+/// ~~~{.cpp}
+/// root[] auto gTA = ROOT::Internal::GetGlobalTaskArena() //get a shared_ptr to the global arena
+/// root[] gTA->InitGlobalTaskArena(nWorkers) // Initialize the global arena and enable Thread Safety in ROOT
```
[the remark this answers is not in the document]
Comment (xvallspl) [reply core/imt/src/RTaskArena.cxx:35]: It used to....fixing it

File: core/imt/inc/ROOT/RTaskArena.hxx (hunk 12 above)
Comment (xvallspl) [reply core/imt/inc/ROOT/RTaskArena.hxx:1]: @Axel-Naumann Is this still needed?

File: core/imt/inc/ROOT/RTaskArena.hxx [hunk 21]
```diff
@@ -0,0 +1,79 @@
...
+////////////////////////////////////////////////////////////////////////////////
+class RTaskArenaWrapper {
+public:
+   RTaskArenaWrapper(unsigned maxConcurrency = 0);
```
[the remark this answers is not in the document]
Comment (xvallspl) [reply core/imt/inc/ROOT/RTaskArena.hxx:55]: Maybe I'm doing something not so nice, but this allows me to call RTaskArena::TaskArenaSize and return fNWorkers when there is no instance of RTaskArena. This was the previous behaviour of TPoolManager and it is useful f.e. when calling GetThreadPoolSize if the task arena is not active.

File: core/imt/src/TImplicitMT.cxx (hunk 17 above)
[the remark this answers is not in the document]
Comment (xvallspl) [reply core/imt/src/TImplicitMT.cxx:48]: Maybe you should ask Danilo or @etejedor  for this one

File: core/imt/src/RTaskArena.cxx (hunk 16 above)
[the remark this answers is not in the document]
Comment (xvallspl) [reply core/imt/src/RTaskArena.cxx:87]: You are absolutely right. This is another remnant from moving the initializer inside the constructor. 🤦

File: core/imt/inc/ROOT/RTaskArena.hxx (hunk 14 above)
Comment (xvallspl) [reply core/imt/inc/ROOT/RTaskArena.hxx:51]: The idea was to have a single task arena. If someone wants a second task arena, they can manipulate tbb directly.

File: core/imt/src/TImplicitMT.cxx (hunk 17 above)
[the remark this answers is not in the document]
Comment (eguiraud) [reply core/imt/src/TImplicitMT.cxx:48]: fair enough, adding this to my pile

File: core/imt/inc/ROOT/RTaskArena.hxx (hunk 14 above)
Comment (eguiraud) [reply core/imt/inc/ROOT/RTaskArena.hxx:51]: @hageboeck `RTaskArenaWrapper` is a singleton, not `tbb::task_arena`.

btw about actually making this a singleton see my comment about changing RTaskArenaWrapper's constructor to private

File: core/imt/inc/ROOT/RTaskArena.hxx (hunk 21 above)
[the remark this answers is not in the document]
Comment (eguiraud) [reply core/imt/inc/ROOT/RTaskArena.hxx:55]: your comment refers to another thread (i'll call it thread 2.) that for some reason github does not make "commentable".

original thread 2. comment:

> static implies you can have multiple RTaskArenaWrapper objects but they must have the same fNWorkers. But by design you can't have multiple RTaskArenaWrappers, so I would argue that static is actually misleading.
>
> can fNWorkers and TaskArenaSize be normal class members?

now about your reply to 2.: so what you actually want is a way to check whether the RTaskArenaWrapper exists or not, which `GetGlobalTaskArena` does not let you do. uhm ok, i don't really have a proposal that is not basically the same thing but different. I will suggest a comment for the static getter.

File: core/imt/inc/ROOT/RTaskArena.hxx [hunk 22]
```diff
@@ -0,0 +1,79 @@
...
+   ~RTaskArenaWrapper(); // necessary to set size back to zero
+   static unsigned TaskArenaSize();
```
Comment (eguiraud) [inline core/imt/inc/ROOT/RTaskArena.hxx:57]:
```suggestion
   static unsigned TaskArenaSize(); //< A static getter that also lets us check for RTaskArenaWrapper's existence 
```

File: core/imt/inc/ROOT/RTaskArena.hxx (hunk 10 above)
[the remark this answers is not in the document]
Comment (eguiraud) [reply core/imt/inc/ROOT/RTaskArena.hxx:61]: more discussion on this topic at https://github.com/root-project/root/pull/5582/files#r437387183

File: core/imt/inc/ROOT/RTaskArena.hxx (hunk 14 above)
Comment (hageboeck) [reply core/imt/inc/ROOT/RTaskArena.hxx:51]: That was my point. It's not a singleton, yet, but it could be. I'm not sure if it *need* be, though.

File: core/imt/inc/ROOT/RTaskArena.hxx (hunk 21 above)
[the remark this answers is not in the document]
Comment (xvallspl) [reply core/imt/inc/ROOT/RTaskArena.hxx:55]: @eguiraud  Even if I friend the factory function what I actually need is to friend make_shared, the way it's programmed right now. Found some workarounds for this (https://stackoverflow.com/a/25069711) but they still leave a public way to call the constructor (in this case through a shared_ptr returning call to the constructor)

That said I could always initialize the shared_ptr without std::make_shared, i.e. `std::shared_ptr<RTaskArena> sp(new RTaskArena(n)` which would introduce one more allocation, but this will only happen "from grapes to pears." Looks like a good solution, am I missing something?

File: core/imt/inc/ROOT/RTaskArena.hxx (hunk 14 above)
Comment (eguiraud) [reply core/imt/inc/ROOT/RTaskArena.hxx:51]: yes, the whole design revolves around ROOT only ever having zero or one RTaskArenaWrappers in flight

File: core/imt/test/testRTaskArena.cxx (hunk 18 above)
[the remark this answers is not in the document]
Comment (hageboeck) [reply core/imt/test/testRTaskArena.cxx:32]: True. Also, it doesn't have side effects, so you can call it as much as you want.

File: core/imt/inc/ROOT/RTaskArena.hxx (hunk 21 above)
[the remark this answers is not in the document]
Comment (eguiraud) [reply core/imt/inc/ROOT/RTaskArena.hxx:55]: `std::shared_ptr<RTaskArena> sp(new RTaskArena(n)` is perfectly fine imo

File: core/imt/test/testRTaskArena.cxx (hunk 19 above)
[the remark this answers is not in the document]
Comment (hageboeck) [reply core/imt/test/testRTaskArena.cxx:78]: Are any of those tests printing warnings? I assume the answer is yes. I want to disallow warnings and errors in the future, so all tests need to capture them. Could you use one of the three methods suggested in #5591 to capture all warnings where you expect them?

[the remark this answers is not in the document]
Comment (xvallspl) [reply core/imt/test/testRTaskArena.cxx:54]: Once it's merged :)

File: core/imt/inc/ROOT/RTaskArena.hxx (hunk 12 above)
Comment (xvallspl) [reply core/imt/inc/ROOT/RTaskArena.hxx:1]: "It is not needed anymore. It's nice to keep it, to signal which module a file (esp a header, think include/) is in."

File: core/imt/test/testRTaskArena.cxx (hunk 19 above)
[the remark this answers is not in the document]
Comment (xvallspl) [reply core/imt/test/testRTaskArena.cxx:54]: Will merge and send a patch for https://github.com/root-project/root/pull/5591

File: core/imt/inc/ROOT/RTaskArena.hxx (hunk 10 above)
[the remark this answers is not in the document]
Comment (eguiraud) [reply core/imt/inc/ROOT/RTaskArena.hxx:73]: I think we can mark this as resolved, we will accept patches with more elegant solutions if they come along

**Comment by xvallspl** [comment]
Sorry, I run clang-format before but I guess I missed a file (?). 

I disagree with clang-format on what's left, for legibility reasons.

**Comment by eguiraud** [comment]
All good. @xvallspl do all commits build (i.e. I can do a rebase and merge), or can I do a squash (i.e. all commits actually amount to one logical change) or do you want to rebase and fix up commits on your side?

**Comment by xvallspl** [comment]
Rebase and merge would be the quickest option. I would not squash them but, If necessary, I can reduce the amount of commits on my side.
